neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/query/parsing/impl/Neo4jQueryRendererDelegate.java

```diff
@@ -71,6 +76,35 @@ private void match(StringBuilder queryBuilder, String targetAlias, String label)
 		node( queryBuilder, targetAlias, label );
 	}
 
+	private void optionalMatch(StringBuilder queryBuilder, String targetAlias) {
```

Comment:
>What part of the query does this method exactly add?

It the OPTIONAL MATCH in the cypher query, the part related to
```  
MATCH (e:WithEmbedded)
OPTIONAL MATCH (e) -[:anEmbeddable]-> (`e_0`:EMBEDDED) -[:anotherEmbeddable]->(`e_0_0`:EMBEDDED)
...
```

> I'm confused a bit by the general organization of the logic. I don't think that the renderer delegate should "know" the resolver delegate and invoke methods from it. Rather the alias handling should be part of the predicates IMO. The predicate factory could assign the right alias names depending on the "depth" of the predicate to be created

There are two places where you can refer to embeddedable at the moment, in the SELECT and in the predicates. I think the part related to the predicates work as you would expect: embedded are added when found via method `Neo4jPredicateFactory#alias(...)`.

The embedded aliases for the select clause get added when they get found via `Neo4jQueryRenderDelegate#setPropertyPath(...)`. I need to check both locations because you could refer to an embedded only in the SELECT clause.

It seems reasonable to me to leave the resolution of the aliases to the  
>"Rendering" here should be more or less the invocation of the predicate tree root node to render itself, which in turn (recursively) asks all children to render themselves (see the call to builder.build() in MongoDBQueryRendererDelegate#getResult(. So e.g. "rendering" a Neo4jComparisonPredicate would yield the complete query part making up that comparison, including the node path in case its an embeddable.